dvc/repo/reproduce.py

```diff
@@ -56,43 +73,38 @@ def reproduce(
             config.SECTION_CORE_INTERACTIVE, False
         )
 
-    targets = []
-    if recursive and os.path.isdir(target):
-        G = self.graph(from_directory=target)[1]
-        dir_targets = [
-            os.path.join(self.root_dir, n) for n in nx.dfs_postorder_nodes(G)
-        ]
-        targets.extend(dir_targets)
-    elif pipeline or all_pipelines:
-        if pipeline:
+    active_graph = _get_active_graph(self.graph)
+    active_pipelines = get_pipelines(active_graph)
+
+    if pipeline or all_pipelines:
+        if all_pipelines:
+            pipelines = active_pipelines
+        else:
             stage = Stage.load(self, target)
             node = relpath(stage.path, self.root_dir)
-            pipelines = [self._get_pipeline(node)]
-        else:
-            pipelines = self.pipelines()
+            pipelines = [get_pipeline(active_pipelines, node)]
 
+        targets = []
         for G in pipelines:
+            attrs = nx.get_node_attributes(G, "stage")
             for node in G.nodes():
                 if G.in_degree(node) == 0:
-                    targets.append(os.path.join(self.root_dir, node))
+                    targets.append(attrs[node])
```

Comment:
This all could be:
```python
targets = [stage for G in pipelines for stage in G if G.in_degree(stage) == 0]
```

Reply:
That is terrible to do such a long comprehensions. Much better to just leave them structured.

Reply:
You can break list comprehension into several lines, this would be the same but more efficient and descriptive. Collecting results with `list.append()` always looked like a smell to me.